### DISCUSSION TITLE
backupMirthServer Code Tempalte

### DISCUSSION THREAD

**Discussion by rjkroll** [nextgenhealthcare/connect#5935] [General]
Hi,

We upgraded to Mirth 4.4 recently and are having issues using the backupMirth server code template that was popular on the other forum.  I know we probably need to update versions, but it looks like other forum is having some issues being up right now. 

Does anyone have an updated version of the code templates involved by chance?

We are getting this in the dashboard log, but it looks like it actually writes the files:

` ERROR  (transformer:138): unable to write file "e:/Mirth_Backups/2023-09-25/MirthDevServer.131_2023-09-25 09-18-33.zip": [JavaPackage org.apache.commons.configuration.PropertiesConfiguration] is not a function, it is object.`

But then in the messages, looking at the error message shows this:

```
CHANNEL:	BackupMirth_Instances
CONNECTOR:	sourceConnector
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
284: 	content =  org.apache.commons.io.FileUtils.listFilesAndDirs(directory, org.apache.commons.io.filefilter.DirectoryFileFilter.DIRECTORY, org.apache.commons.io.filefilter.DirectoryFileFilter.DIRECTORY);	
285: 	// remove the parent dir from the result list
286: 	content.remove(directory);
287: 	// and transform the collection to an array
288: 	content = content.toArray();
289: 	logger.debug('Found ' + content.length  + ' subfolder in directory (' + content[1].getPath() + ')');
290: 	// sort list of fetched directories having the youngest on top (which should be )
291: 	java.util.Arrays.sort(content, org.apache.commons.io.comparator.LastModifiedFileComparator.LASTMODIFIED_REVERSE);
292: 	// calculate the threshold date
293: 	var dateThreshold =  new Date();
LINE NUMBER:	289
DETAILS:	TypeError: Cannot call method "getPath" of undefined
```

I've tried commenting out some things related to these errors related to getPath and they are no longer producing error messages that are viewable as ERRORS, they now show SENT, but the error in the dashboard log saying 'unable to write' still exists even though I see a ZIP file with the backup XML in it.  So, just trying to see if the template has been updated recently to work with 4.0+ or higher.

Thanks,

**Comment by pacmano1** [comment]
@odoodo wrote that.

**Reply by rjkroll** [reply]
Thanks.   Hopefully he sees this and replies, as its a very useful template to have.   Wish it was a built-in feature in Mirth to automate backups like this instead of having to rely on other solutions.

**Reply by pacmano1** [reply]
Create a feature reqeust in the issues section.

**Comment by odoodo** [comment]
It seems like it writes the backup of the mirth configuration but it does not add the configuration map to the zip file. In later versions of Mirth Apache Commons lib has been replaced by Apache Commons 2. The code that seems to fail in your case is the part that detects the version of the Commons lib. 

As we are currently still on v3.7.1 (will hopefully change around the end of the year) it was not possible for me to test when I added the support for later Mirth versions. Thus, working blindfolded, it seems like I forgot to instantiate the PropertiesConfiguration for commons v2 (**new** after ? is missing). 
```js
// prepare structure
var properties = useCommons2 ? org.apache.commons.configuration2.PropertiesConfiguration() : new org.apache.commons.configuration.PropertiesConfiguration();
```
(I actually wonder why nobody reported an issue before ;-) )

You can find the fixed code template [here](https://forums.mirthproject.io/forum/mirth-connect/support/16298-code-template-automatic-mirth-backup#post16298).  Please let me know if this worked for you.

**Comment by rjkroll** [comment]
Thanks for the help.   Updating to the new version of the code template seems to have resolve the issues.
